File: spike/cpp/src/storage.cpp
#include "storage.h"

#include <atomic>
#include <chrono>

#include "rocksdb/iterator.h"
#include "rocksdb/options.h"
#include "rocksdb/write_batch.h"

namespace gamestore {

uint64_t now_ms() {
  using namespace std::chrono;
  return duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
}

static uint64_t now_micros() {
  using namespace std::chrono;
  return duration_cast<microseconds>(system_clock::now().time_since_epoch()).count();
}

// Monotonic, globally increasing structure version (cf. HLC in the design).
static std::atomic<uint64_t> g_version_clock{0};

static uint64_t next_version() {
  for (;;) {
    uint64_t prev = g_version_clock.load(std::memory_order_seq_cst);
    uint64_t candidate = std::max(prev + 1, now_micros());
    if (g_version_clock.compare_exchange_strong(prev, candidate)) {
      return candidate;
    }
  }
}

std::unique_ptr<Store> Store::Open(const std::string& path) {
  auto store = std::unique_ptr<Store>(new Store());
  store->versions_ = std::make_unique<VersionMap>();
  store->filter_ = std::make_unique<SubkeyGcFilter>(store->versions_.get());

  rocksdb::Options options;
  options.create_if_missing = true;
  options.compaction_filter = store->filter_.get();

  rocksdb::DB* db = nullptr;
  rocksdb::Status s = rocksdb::DB::Open(options, path, &db);
  if (!s.ok()) {
    fprintf(stderr, "[cpp] failed to open RocksDB at %s: %s\n", path.c_str(),
            s.ToString().c_str());
    return nullptr;
  }
  store->db_ = db;
  store->RebuildVersionMap();
  return store;
}

Store::~Store() {
  delete db_;
}

void Store::RebuildVersionMap() {
  std::unique_ptr<rocksdb::Iterator> it(db_->NewIterator(rocksdb::ReadOptions()));
  for (it->SeekToFirst(); it->Valid(); it->Next()) {
    const rocksdb::Slice k = it->key();
    if (k.size() >= 1 && static_cast<uint8_t>(k.data()[0]) == META_PREFIX) {
      auto meta = Meta::decode(it->value().ToString());
      if (meta && meta->type_id == TYPE_HASH) {
        versions_->set(std::string(k.data() + 1, k.size() - 1), meta->version);
      }
    }
  }
}

bool Store::LoadMeta(const std::string& key, Meta* out) {
  std::string raw;
  rocksdb::Status s = db_->Get(rocksdb::ReadOptions(), meta_key(key), &raw);
  if (!s.ok()) {
    return false;
  }
  auto meta = Meta::decode(raw);
  if (!meta) {
    return false;
  }
  *out = std::move(*meta);
  return true;
}

bool Store::LoadLiveMeta(const std::string& key, Meta* out) {
  Meta meta;
  if (!LoadMeta(key, &meta)) {
    return false;
  }
  if (meta.expire_ms != 0 && now_ms() >= meta.expire_ms) {
    LogicalDelete(key);
    return false;
  }
  *out = std::move(meta);
  return true;
}

void Store::PutMeta(const std::string& key, const Meta& meta) {
  db_->Put(rocksdb::WriteOptions(), meta_key(key), meta.encode());
}

void Store::LogicalDelete(const std::string& key) {
  db_->Delete(rocksdb::WriteOptions(), meta_key(key));
  versions_->remove(key);
}

// ---- String ----------------------------------------------------------------

void Store::Set(const std::string& key, const std::string& value, uint64_t expire_ms) {
  versions_->remove(key);  // overwriting drops any prior structure
  Meta meta;
  meta.type_id = TYPE_STRING;
  meta.version = next_version();
  meta.expire_ms = expire_ms;
  meta.payload = value;
  PutMeta(key, meta);
}

bool Store::Get(const std::string& key, std::string* out) {
  Meta meta;
  if (!LoadLiveMeta(key, &meta) || meta.type_id != TYPE_STRING) {
    return false;
  }
  *out = meta.payload;
  return true;
}

// ---- Generic ----------------------------------------------------------------

bool Store::Exists(const std::string& key) {
  Meta meta;
  return LoadLiveMeta(key, &meta);
}

const char* Store::TypeOf(const std::string& key) {
  Meta meta;
  if (!LoadLiveMeta(key, &meta)) {
    return "none";
  }
  if (meta.type_id == TYPE_STRING) return "string";
  if (meta.type_id == TYPE_HASH) return "hash";
  return "none";
}

bool Store::Del(const std::string& key) {
  Meta meta;
  if (!LoadLiveMeta(key, &meta)) {
    return false;
  }
  LogicalDelete(key);
  return true;
}

int64_t Store::ExpireMs(const std::string& key, uint64_t expire_at_ms) {
  Meta meta;
  if (!LoadLiveMeta(key, &meta)) {
    return 0;
  }
  meta.expire_ms = expire_at_ms;
  PutMeta(key, meta);
  return 1;
}

int64_t Store::Pttl(const std::string& key) {
  Meta meta;
  if (!LoadLiveMeta(key, &meta)) {
    return -2;
  }
  if (meta.expire_ms == 0) {
    return -1;
  }
  uint64_t now = now_ms();
  return meta.expire_ms > now ? static_cast<int64_t>(meta.expire_ms - now) : 0;
}

// ---- Hash -------------------------------------------------------------------

int64_t Store::Hset(const std::string& key,
                    const std::vector<std::pair<std::string, std::string>>& pairs) {
  Meta meta;
  bool have = LoadLiveMeta(key, &meta) && meta.type_id == TYPE_HASH;
  if (!have) {
    meta = Meta();
    meta.type_id = TYPE_HASH;
    meta.version = next_version();
    meta.expire_ms = 0;
    meta.set_field_count(0);
    versions_->set(key, meta.version);
  }

  uint32_t field_count = meta.field_count();
  int64_t created = 0;
  rocksdb::WriteBatch batch;
  for (const auto& [field, value] : pairs) {
    std::string sk = subkey(key, meta.version, field);
    std::string existing;
    bool existed = db_->Get(rocksdb::ReadOptions(), sk, &existing).ok();
    if (!existed) {
      field_count += 1;
      created += 1;
    }
    batch.Put(sk, value);
  }
  meta.set_field_count(field_count);
  batch.Put(meta_key(key), meta.encode());
  db_->Write(rocksdb::WriteOptions(), &batch);
  return created;
}

bool Store::Hget(const std::string& key, const std::string& field, std::string* out) {
  Meta meta;
  if (!LoadLiveMeta(key, &meta) || meta.type_id != TYPE_HASH) {
    return false;
  }
  return db_->Get(rocksdb::ReadOptions(), subkey(key, meta.version, field), out).ok();
}

std::vector<std::pair<std::string, std::string>> Store::Hgetall(const std::string& key) {
  std::vector<std::pair<std::string, std::string>> out;
  Meta meta;
  if (!LoadLiveMeta(key, &meta) || meta.type_id != TYPE_HASH) {
    return out;
  }
  std::string prefix = subkey_prefix(key, meta.version);
  std::unique_ptr<rocksdb::Iterator> it(db_->NewIterator(rocksdb::ReadOptions()));
  for (it->Seek(prefix); it->Valid(); it->Next()) {
    const rocksdb::Slice k = it->key();
    if (k.size() < prefix.size() || memcmp(k.data(), prefix.data(), prefix.size()) != 0) {
      break;
    }
    std::string field(k.data() + prefix.size(), k.size() - prefix.size());
    out.emplace_back(std::move(field), it->value().ToString());
  }
  return out;
}

int64_t Store::Hdel(const std::string& key, const std::vector<std::string>& fields) {
  Meta meta;
  if (!LoadLiveMeta(key, &meta) || meta.type_id != TYPE_HASH) {
    return 0;
  }
  uint32_t field_count = meta.field_count();
  int64_t removed = 0;
  rocksdb::WriteBatch batch;
  for (const auto& field : fields) {
    std::string sk = subkey(key, meta.version, field);
    std::string existing;
    if (db_->Get(rocksdb::ReadOptions(), sk, &existing).ok()) {
      batch.Delete(sk);
      removed += 1;
      if (field_count > 0) field_count -= 1;
    }
  }
  if (removed > 0) {
    if (field_count == 0) {
      LogicalDelete(key);
      db_->Write(rocksdb::WriteOptions(), &batch);
    } else {
      meta.set_field_count(field_count);
      batch.Put(meta_key(key), meta.encode());
      db_->Write(rocksdb::WriteOptions(), &batch);
    }
  }
  return removed;
}

int64_t Store::Hlen(const std::string& key) {
  Meta meta;
  if (!LoadLiveMeta(key, &meta) || meta.type_id != TYPE_HASH) {
    return 0;
  }
  return meta.field_count();
}

bool Store::Hexists(const std::string& key, const std::string& field) {
  std::string tmp;
  return Hget(key, field, &tmp);
}

// ---- Admin / introspection --------------------------------------------------

void Store::FlushDb() {
  rocksdb::WriteBatch batch;
  std::unique_ptr<rocksdb::Iterator> it(db_->NewIterator(rocksdb::ReadOptions()));
  for (it->SeekToFirst(); it->Valid(); it->Next()) {
    batch.Delete(it->key());
  }
  db_->Write(rocksdb::WriteOptions(), &batch);
  versions_->clear();
}

int64_t Store::DbSize() {
  int64_t n = 0;
  std::unique_ptr<rocksdb::Iterator> it(db_->NewIterator(rocksdb::ReadOptions()));
  for (it->SeekToFirst(); it->Valid(); it->Next()) {
    const rocksdb::Slice k = it->key();
    if (k.size() >= 1 && static_cast<uint8_t>(k.data()[0]) == META_PREFIX) {
      n += 1;
    }
  }
  return n;
}

int64_t Store::RawSubkeyCount() {
  int64_t n = 0;
  std::unique_ptr<rocksdb::Iterator> it(db_->NewIterator(rocksdb::ReadOptions()));
  for (it->SeekToFirst(); it->Valid(); it->Next()) {
    const rocksdb::Slice k = it->key();
    if (k.size() >= 1 && static_cast<uint8_t>(k.data()[0]) == SUBKEY_PREFIX) {
      n += 1;
    }
  }
  return n;
}

void Store::Compact() {
  // Flush first (the filter only sees SST data, not the memtable) and force the
  // bottommost level so RocksDB cannot skip the rewrite via trivial move.
  db_->Flush(rocksdb::FlushOptions());
  rocksdb::CompactRangeOptions cro;
  cro.bottommost_level_compaction = rocksdb::BottommostLevelCompaction::kForce;
  db_->CompactRange(cro, nullptr, nullptr);
}

}  // namespace gamestore

File: spike/rust/src/storage.rs
//! Storage layer: Redis data model encoded onto RocksDB (the "general engine
//! layer" in `docs/design/03-storage-engine.md`).

use std::sync::atomic::{AtomicU64, Ordering};
use std::time::{SystemTime, UNIX_EPOCH};

use rocksdb::{
    compaction_filter::Decision, DBCompactionStyle, IteratorMode, Options, WriteBatch, DB,
};

use crate::encoding::{self, Meta, TYPE_HASH, TYPE_STRING};
use crate::gc::VersionMap;

pub fn now_ms() -> u64 {
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as u64
}

fn now_micros() -> u64 {
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_micros() as u64
}

/// Monotonic, globally increasing structure version. Seeded from wall-clock
/// microseconds so versions keep increasing across restarts and a deleted key
/// can never collide with a freshly rebuilt one (cf. HLC in the design docs).
static VERSION_CLOCK: AtomicU64 = AtomicU64::new(0);

fn next_version() -> u64 {
    loop {
        let prev = VERSION_CLOCK.load(Ordering::SeqCst);
        let candidate = std::cmp::max(prev + 1, now_micros());
        if VERSION_CLOCK
            .compare_exchange(prev, candidate, Ordering::SeqCst, Ordering::SeqCst)
            .is_ok()
        {
            return candidate;
        }
    }
}

pub struct Store {
    db: DB,
    versions: VersionMap,
}

impl Store {
    pub fn open(path: &str) -> Result<Store, String> {
        let versions = VersionMap::new();

        let mut opts = Options::default();
        opts.create_if_missing(true);
        opts.set_compaction_style(DBCompactionStyle::Level);
        // Install the version-based subkey GC as a compaction filter.
        let vm = versions.clone();
        opts.set_compaction_filter("gamestore-subkey-gc", move |_level, key, _value| {
            if vm.should_keep(key) {
                Decision::Keep
            } else {
                Decision::Remove
            }
        });

        let db = DB::open(&opts, path).map_err(|e| e.to_string())?;
        let store = Store { db, versions };
        store.rebuild_version_map();
        Ok(store)
    }

    /// Repopulate the in-memory version map from persisted metadata on startup.
    fn rebuild_version_map(&self) {
        let iter = self.db.iterator(IteratorMode::Start);
        for item in iter {
            let (k, v) = match item {
                Ok(kv) => kv,
                Err(_) => continue,
            };
            if k.first() == Some(&encoding::META_PREFIX) {
                if let Some(meta) = Meta::decode(&v) {
                    if meta.type_id == TYPE_HASH {
                        let user_key = &k[1..];
                        self.versions.set(user_key, meta.version);
                    }
                }
            }
        }
    }

    fn load_meta(&self, user_key: &[u8]) -> Option<Meta> {
        let raw = self.db.get(encoding::meta_key(user_key)).ok()??;
        Meta::decode(&raw)
    }

    /// Load metadata, applying lazy expiration: an expired key is treated as
    /// missing and physically logical-deleted (version dropped from the map so
    /// its subkeys become collectable).
    fn load_live_meta(&self, user_key: &[u8]) -> Option<Meta> {
        let meta = self.load_meta(user_key)?;
        if meta.expire_ms != 0 && now_ms() >= meta.expire_ms {
            self.logical_delete(user_key);
            return None;
        }
        Some(meta)
    }

    fn put_meta(&self, user_key: &[u8], meta: &Meta) {
        self.db
            .put(encoding::meta_key(user_key), meta.encode())
            .expect("put meta");
    }

    fn logical_delete(&self, user_key: &[u8]) {
        // Remove metadata + drop from version map. Subkeys (if any) are now
        // orphaned and will be reclaimed by the compaction filter.
        let _ = self.db.delete(encoding::meta_key(user_key));
        self.versions.remove(user_key);
    }

    // ---- String ----------------------------------------------------------

    pub fn set(&self, key: &[u8], value: &[u8], expire_ms: u64) {
        // Overwriting with a String drops any previous (e.g. Hash) structure.
        self.versions.remove(key);
        let meta = Meta {
            type_id: TYPE_STRING,
            version: next_version(),
            expire_ms,
            payload: value.to_vec(),
        };
        self.put_meta(key, &meta);
    }

    pub fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
        let meta = self.load_live_meta(key)?;
        if meta.type_id != TYPE_STRING {
            return None;
        }
        Some(meta.payload)
    }

    // ---- Generic ----------------------------------------------------------

    pub fn exists(&self, key: &[u8]) -> bool {
        self.load_live_meta(key).is_some()
    }

    pub fn type_of(&self, key: &[u8]) -> &'static str {
        match self.load_live_meta(key) {
            Some(m) if m.type_id == TYPE_STRING => "string",
            Some(m) if m.type_id == TYPE_HASH => "hash",
            _ => "none",
        }
    }

    pub fn del(&self, key: &[u8]) -> bool {
        if self.load_live_meta(key).is_some() {
            self.logical_delete(key);
            true
        } else {
            false
        }
    }

    /// Returns 1 if a timeout was set, 0 if the key does not exist.
    pub fn expire_ms(&self, key: &[u8], expire_at_ms: u64) -> i64 {
        match self.load_live_meta(key) {
            Some(mut meta) => {
                meta.expire_ms = expire_at_ms;
                self.put_meta(key, &meta);
                1
            }
            None => 0,
        }
    }

    /// -2 if no key, -1 if no expire, else remaining milliseconds.
    pub fn pttl(&self, key: &[u8]) -> i64 {
        match self.load_live_meta(key) {
            None => -2,
            Some(meta) => {
                if meta.expire_ms == 0 {
                    -1
                } else {
                    (meta.expire_ms.saturating_sub(now_ms())) as i64
                }
            }
        }
    }

    // ---- Hash -------------------------------------------------------------

    /// Returns the number of *new* fields created.
    pub fn hset(&self, key: &[u8], pairs: &[(Vec<u8>, Vec<u8>)]) -> i64 {
        let mut meta = match self.load_live_meta(key) {
            Some(m) if m.type_id == TYPE_HASH => m,
            _ => {
                // New (or replacing a non-hash) -> fresh version.
                let version = next_version();
                let mut m = Meta {
                    type_id: TYPE_HASH,
                    version,
                    expire_ms: 0,
                    payload: Vec::new(),
                };
                m.set_field_count(0);
                self.versions.set(key, version);
                m
            }
        };

        let mut field_count = meta.field_count();
        let mut created = 0i64;
        let mut batch = WriteBatch::default();
        for (field, value) in pairs {
            let sk = encoding::subkey(key, meta.version, field);
            let existed = self.db.get(&sk).ok().flatten().is_some();
            if !existed {
                field_count += 1;
                created += 1;
            }
            batch.put(&sk, value);
        }
        meta.set_field_count(field_count);
        batch.put(encoding::meta_key(key), meta.encode());
        self.db.write(batch).expect("hset batch");
        created
    }

    pub fn hget(&self, key: &[u8], field: &[u8]) -> Option<Vec<u8>> {
        let meta = self.load_live_meta(key)?;
        if meta.type_id != TYPE_HASH {
            return None;
        }
        let sk = encoding::subkey(key, meta.version, field);
        self.db.get(sk).ok().flatten()
    }

    pub fn hgetall(&self, key: &[u8]) -> Vec<(Vec<u8>, Vec<u8>)> {
        let meta = match self.load_live_meta(key) {
            Some(m) if m.type_id == TYPE_HASH => m,
            _ => return Vec::new(),
        };
        let prefix = encoding::subkey_prefix(key, meta.version);
        let mut out = Vec::new();
        let iter = self
            .db
            .iterator(IteratorMode::From(&prefix, rocksdb::Direction::Forward));
        for item in iter {
            let (k, v) = match item {
                Ok(kv) => kv,
                Err(_) => break,
            };
            if !k.starts_with(&prefix) {
                break;
            }
            let field = k[prefix.len()..].to_vec();
            out.push((field, v.to_vec()));
        }
        out
    }

    pub fn hdel(&self, key: &[u8], fields: &[Vec<u8>]) -> i64 {
        let mut meta = match self.load_live_meta(key) {
            Some(m) if m.type_id == TYPE_HASH => m,
            _ => return 0,
        };
        let mut field_count = meta.field_count();
        let mut removed = 0i64;
        let mut batch = WriteBatch::default();
        for field in fields {
            let sk = encoding::subkey(key, meta.version, field);
            if self.db.get(&sk).ok().flatten().is_some() {
                batch.delete(&sk);
                removed += 1;
                field_count = field_count.saturating_sub(1);
            }
        }
        if removed > 0 {
            if field_count == 0 {
                self.logical_delete(key);
                self.db.write(batch).expect("hdel batch");
            } else {
                meta.set_field_count(field_count);
                batch.put(encoding::meta_key(key), meta.encode());
                self.db.write(batch).expect("hdel batch");
            }
        }
        removed
    }

    pub fn hlen(&self, key: &[u8]) -> i64 {
        match self.load_live_meta(key) {
            Some(m) if m.type_id == TYPE_HASH => m.field_count() as i64,
            _ => 0,
        }
    }

    pub fn hexists(&self, key: &[u8], field: &[u8]) -> bool {
        self.hget(key, field).is_some()
    }

    // ---- Admin / introspection (spike-only) -------------------------------

    pub fn flushdb(&self) {
        let keys: Vec<Box<[u8]>> = self
            .db
            .iterator(IteratorMode::Start)
            .filter_map(|item| item.ok().map(|(k, _)| k))
            .collect();
        let mut batch = WriteBatch::default();
        for k in &keys {
            batch.delete(k);
        }
        self.db.write(batch).expect("flush batch");
        self.versions_clear();
    }

    fn versions_clear(&self) {
        // Rebuild as empty by removing every tracked key.
        // Cheaper: just drop entries we know about via a fresh scan.
        let keys: Vec<Box<[u8]>> = self
            .db
            .iterator(IteratorMode::Start)
            .filter_map(|item| item.ok().map(|(k, _)| k))
            .collect();
        for k in keys {
            if k.first() == Some(&encoding::META_PREFIX) {
                self.versions.remove(&k[1..]);
            }
        }
    }

    /// Total number of physical metadata records (logical key count).
    pub fn dbsize(&self) -> i64 {
        self.db
            .iterator(IteratorMode::Start)
            .filter_map(|i| i.ok())
            .filter(|(k, _)| k.first() == Some(&encoding::META_PREFIX))
            .count() as i64
    }

    /// Total number of physical subkey records currently stored. Used by the
    /// functional test to prove the compaction filter reclaimed stale subkeys.
    pub fn raw_subkey_count(&self) -> i64 {
        self.db
            .iterator(IteratorMode::Start)
            .filter_map(|i| i.ok())
            .filter(|(k, _)| k.first() == Some(&encoding::SUBKEY_PREFIX))
            .count() as i64
    }

    /// Force a full compaction so the compaction filter runs synchronously.
    /// Flush first (the filter only sees SST data, not the memtable) and force
    /// the bottommost level so RocksDB cannot skip the rewrite via trivial move.
    pub fn compact(&self) {
        let _ = self.db.flush();
        let mut copts = rocksdb::CompactOptions::default();
        copts.set_bottommost_level_compaction(rocksdb::BottommostLevelCompaction::Force);
        self.db
            .compact_range_opt::<&[u8], &[u8]>(None, None, &copts);
    }
}
